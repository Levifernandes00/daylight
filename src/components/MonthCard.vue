<template>
  <main class="bg-white shadow-1 border-round flex-grow-1 p-2 m-1">
    <h3>{{ date }}</h3>
    <div class="px-6 py-3">
      <div class="py-1">
        <small>Day light:</small>
        <h2>{{ daylight }}</h2>
      </div>
    </div>
  </main>
</template>

<script lang="ts">
import { useStore } from '@/store/index'
import { DateTime } from 'luxon'
import { defineComponent } from 'vue'

export default defineComponent({
  props: {
    month: String
  },
  data() {
    return {
      date: '--',
      daylight: '--'
    }
  },
  async created() {
    const today = DateTime.now()
    const date = DateTime.utc(today.year, parseInt(this.month!), today.day)
    this.setDate(date)
    this.setDaylight(date)
  },
  methods: {
    setDate(date: DateTime) {
      this.date = date
        .setLocale('it-IT')
        .toLocaleString(DateTime.DATE_MED_WITH_WEEKDAY)
    },
    setDaylight(date: DateTime) {
      const store = useStore()
      store.watch(
        () => store.state.location,
        async () => {
          this.daylight = await store.dispatch('fetchDaylight', date)
        }
      )
    }
  },
  mounted() {
    this.month
  }
})
</script>

<style lang="scss"></style>
