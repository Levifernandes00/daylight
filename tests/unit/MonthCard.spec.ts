import MonthCard from '@/components/MonthCard.vue'
import { flushPromises, shallowMount } from '@vue/test-utils'
import { DateTime } from 'luxon'

const position = { latitude: 45.6141484, longitude: 8.777796 }

const expectedDaylight = '12:18:32'

const date = DateTime.now()

const expectedDateString = date
.setLocale('it-IT')
.toLocaleString(DateTime.DATE_MED_WITH_WEEKDAY)


jest.mock('@/store', () => ({
  ...jest.fn(),
  useStore: () => ({
    state: {location: position },
    dispatch: (functionName: string, payload: DateTime) => expectedDaylight,
    watch: (getter: () => void, callback: () => void) => {
      getter(), callback()
    }
  })
}))

describe('MonthCard.vue ', () => {
  it('matches snapshot',async () => {
    const wrapper = shallowMount(MonthCard, {props: {cardDate: date}})
    await flushPromises()
    expect(wrapper).toMatchSnapshot()
  })

  it('renders right day', async () => {
    const wrapper = shallowMount(MonthCard, {props: {cardDate: date}})
    await flushPromises()
    expect(wrapper.get('#date').text()).toMatch(expectedDateString)
  })

  it('renders small title', async () => {
    const wrapper = shallowMount(MonthCard, {props: {cardDate: date}})
    expect(wrapper.get('small').text()).toMatch('Day light:')
  })

  it('renders proper daylight', async () => {
    const wrapper = shallowMount(MonthCard, {props: {cardDate: date}})
    await flushPromises()
    expect(wrapper.get('#daylight').text()).toMatch(expectedDaylight)
  })
})
