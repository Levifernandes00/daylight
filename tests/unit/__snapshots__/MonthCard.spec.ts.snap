// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MonthCard.vue  matches snapshot 1`] = `
VueWrapper {
  "__app": Object {
    "_component": Object {
      "name": "VTU_ROOT",
      "render": [Function],
    },
    "_container": <div
      data-v-app=""
    >
      <main
        class="bg-white shadow-1 border-round flex-grow-1 p-2 m-1"
      >
        <h3
          id="date"
        >
          gio 9 giu 2022
        </h3>
        <div
          class="px-6 py-3"
        >
          <div
            class="py-1"
          >
            <small>
              Day light:
            </small>
            <h2
              id="daylight"
            >
              12:18:32
            </h2>
          </div>
        </div>
      </main>
    </div>,
    "_context": Object {
      "app": [Circular],
      "components": Object {
        "transition": Object {
          "name": "transition",
        },
        "transition-group": Object {
          "name": "transition-group",
        },
      },
      "config": Object {
        "compilerOptions": Object {},
        "errorHandler": undefined,
        "globalProperties": Object {},
        "isNativeTag": [Function],
        "optionMergeStrategies": Object {},
        "performance": false,
        "warnHandler": undefined,
      },
      "directives": Object {},
      "emitsCache": WeakMap {},
      "mixins": Array [
        Object {
          "beforeCreate": [Function],
        },
      ],
      "optionsCache": WeakMap {},
      "propsCache": WeakMap {},
      "provides": Object {},
      "reload": [Function],
    },
    "_instance": Object {
      "a": null,
      "accessCache": Object {},
      "appContext": Object {
        "app": [Circular],
        "components": Object {
          "transition": Object {
            "name": "transition",
          },
          "transition-group": Object {
            "name": "transition-group",
          },
        },
        "config": Object {
          "compilerOptions": Object {},
          "errorHandler": undefined,
          "globalProperties": Object {},
          "isNativeTag": [Function],
          "optionMergeStrategies": Object {},
          "performance": false,
          "warnHandler": undefined,
        },
        "directives": Object {},
        "emitsCache": WeakMap {},
        "mixins": Array [
          Object {
            "beforeCreate": [Function],
          },
        ],
        "optionsCache": WeakMap {},
        "propsCache": WeakMap {},
        "provides": Object {},
        "reload": [Function],
      },
      "asyncDep": null,
      "asyncResolved": false,
      "attrs": Object {},
      "bc": null,
      "bm": null,
      "bu": null,
      "bum": null,
      "c": null,
      "components": null,
      "ctx": Object {},
      "da": null,
      "data": Object {},
      "directives": null,
      "ec": null,
      "effect": ReactiveEffect {
        "active": true,
        "allowRecurse": true,
        "deps": Array [
          Set {
            [Circular],
          },
          Set {
            [Circular],
          },
          Set {
            [Circular],
          },
          Set {
            [Circular],
          },
        ],
        "fn": [Function],
        "onTrack": undefined,
        "onTrigger": undefined,
        "parent": undefined,
        "scheduler": [Function],
      },
      "emit": [Function],
      "emitsOptions": null,
      "emitted": null,
      "exposeProxy": null,
      "exposed": null,
      "inheritAttrs": undefined,
      "isDeactivated": false,
      "isMounted": true,
      "isUnmounted": false,
      "m": null,
      "next": null,
      "parent": null,
      "props": Object {},
      "propsDefaults": Object {},
      "propsOptions": Array [
        Object {},
        Array [],
      ],
      "provides": Object {},
      "proxy": Object {},
      "refs": Object {
        "VTU_COMPONENT": Object {
          "cardDate": "2022-06-09T00:00:00.000Z",
          "date": "gio 9 giu 2022",
          "daylight": "12:18:32",
          "hasOwnProperty": [Function],
          "setDate": [Function],
          "setDaylight": [Function],
        },
      },
      "render": [Function],
      "renderCache": Array [],
      "root": [Circular],
      "rtc": null,
      "rtg": null,
      "scope": EffectScope {
        "active": true,
        "cleanups": Array [],
        "effects": Array [
          ReactiveEffect {
            "active": true,
            "allowRecurse": true,
            "deps": Array [
              Set {
                [Circular],
              },
              Set {
                [Circular],
              },
              Set {
                [Circular],
              },
              Set {
                [Circular],
              },
            ],
            "fn": [Function],
            "onTrack": undefined,
            "onTrigger": undefined,
            "parent": undefined,
            "scheduler": [Function],
          },
        ],
      },
      "setupContext": null,
      "setupState": Object {},
      "slots": Object {},
      "sp": null,
      "subTree": Object {
        "__v_isVNode": true,
        "__v_skip": true,
        "anchor": null,
        "appContext": null,
        "children": null,
        "component": Object {
          "a": null,
          "accessCache": Object {
            "@@__IMMUTABLE_ITERABLE__@@": 0,
            "@@__IMMUTABLE_RECORD__@@": 0,
            "_isMockFunction": 0,
            "cardDate": 3,
            "constructor": 0,
            "date": 2,
            "daylight": 2,
            "hasOwnProperty": 4,
            "isVueInstance": 0,
            "setDate": 4,
            "setDaylight": 4,
            "toJSON": 0,
            Symbol(Symbol.toStringTag): 0,
          },
          "appContext": Object {
            "app": [Circular],
            "components": Object {
              "transition": Object {
                "name": "transition",
              },
              "transition-group": Object {
                "name": "transition-group",
              },
            },
            "config": Object {
              "compilerOptions": Object {},
              "errorHandler": undefined,
              "globalProperties": Object {},
              "isNativeTag": [Function],
              "optionMergeStrategies": Object {},
              "performance": false,
              "warnHandler": undefined,
            },
            "directives": Object {},
            "emitsCache": WeakMap {},
            "mixins": Array [
              Object {
                "beforeCreate": [Function],
              },
            ],
            "optionsCache": WeakMap {},
            "propsCache": WeakMap {},
            "provides": Object {},
            "reload": [Function],
          },
          "asyncDep": null,
          "asyncResolved": false,
          "attrs": Object {},
          "bc": null,
          "bm": null,
          "bu": null,
          "bum": null,
          "c": null,
          "components": Object {},
          "ctx": Object {
            "cardDate": "2022-06-09T00:00:00.000Z",
            "date": "gio 9 giu 2022",
            "daylight": "12:18:32",
            "hasOwnProperty": [Function],
            "setDate": [Function],
            "setDaylight": [Function],
          },
          "da": null,
          "data": Object {
            "date": "gio 9 giu 2022",
            "daylight": "12:18:32",
          },
          "directives": null,
          "ec": null,
          "effect": ReactiveEffect {
            "active": true,
            "allowRecurse": true,
            "deps": Array [
              Set {
                [Circular],
              },
              Set {
                [Circular],
              },
            ],
            "fn": [Function],
            "onTrack": undefined,
            "onTrigger": undefined,
            "parent": undefined,
            "scheduler": [Function],
          },
          "emit": [Function],
          "emitsOptions": null,
          "emitted": null,
          "exposeProxy": null,
          "exposed": null,
          "inheritAttrs": undefined,
          "isDeactivated": false,
          "isMounted": true,
          "isUnmounted": false,
          "m": null,
          "next": null,
          "parent": [Circular],
          "props": Object {
            "cardDate": "2022-06-09T00:00:00.000Z",
          },
          "propsDefaults": Object {},
          "propsOptions": Array [
            Object {
              "cardDate": Object {
                "0": false,
                "1": true,
                "type": [Function],
              },
            },
            Array [],
          ],
          "provides": Object {},
          "proxy": Object {
            "cardDate": "2022-06-09T00:00:00.000Z",
            "date": "gio 9 giu 2022",
            "daylight": "12:18:32",
            "hasOwnProperty": [Function],
            "setDate": [Function],
            "setDaylight": [Function],
          },
          "refs": Object {},
          "render": [Function],
          "renderCache": Array [],
          "root": [Circular],
          "rtc": null,
          "rtg": null,
          "scope": EffectScope {
            "active": true,
            "cleanups": Array [],
            "effects": Array [
              ReactiveEffect {
                "active": true,
                "allowRecurse": true,
                "deps": Array [
                  Set {
                    [Circular],
                  },
                  Set {
                    [Circular],
                  },
                ],
                "fn": [Function],
                "onTrack": undefined,
                "onTrigger": undefined,
                "parent": undefined,
                "scheduler": [Function],
              },
            ],
          },
          "setupContext": null,
          "setupState": Object {},
          "slots": Object {},
          "sp": null,
          "subTree": Object {
            "__v_isVNode": true,
            "__v_skip": true,
            "anchor": null,
            "appContext": null,
            "children": Array [
              Object {
                "__v_isVNode": true,
                "__v_skip": true,
                "anchor": null,
                "appContext": null,
                "children": "gio 9 giu 2022",
                "component": null,
                "dirs": null,
                "dynamicChildren": null,
                "dynamicProps": null,
                "el": <h3
                  id="date"
                >
                  gio 9 giu 2022
                </h3>,
                "key": null,
                "patchFlag": 1,
                "props": Object {
                  "id": "date",
                },
                "ref": null,
                "scopeId": null,
                "shapeFlag": 9,
                "slotScopeIds": null,
                "ssContent": null,
                "ssFallback": null,
                "staticCount": 0,
                "suspense": null,
                "target": null,
                "targetAnchor": null,
                "transition": null,
                "type": "h3",
              },
              Object {
                "__v_isVNode": true,
                "__v_skip": true,
                "anchor": null,
                "appContext": null,
                "children": Array [
                  Object {
                    "__v_isVNode": true,
                    "__v_skip": true,
                    "anchor": null,
                    "appContext": null,
                    "children": Array [
                      Object {
                        "__v_isVNode": true,
                        "__v_skip": true,
                        "anchor": null,
                        "appContext": null,
                        "children": "Day light:",
                        "component": null,
                        "dirs": null,
                        "dynamicChildren": null,
                        "dynamicProps": null,
                        "el": <small>
                          Day light:
                        </small>,
                        "key": null,
                        "patchFlag": -1,
                        "props": null,
                        "ref": null,
                        "scopeId": null,
                        "shapeFlag": 9,
                        "slotScopeIds": null,
                        "ssContent": null,
                        "ssFallback": null,
                        "staticCount": 0,
                        "suspense": null,
                        "target": null,
                        "targetAnchor": null,
                        "transition": null,
                        "type": "small",
                      },
                      Object {
                        "__v_isVNode": true,
                        "__v_skip": true,
                        "anchor": null,
                        "appContext": null,
                        "children": "12:18:32",
                        "component": null,
                        "dirs": null,
                        "dynamicChildren": null,
                        "dynamicProps": null,
                        "el": <h2
                          id="daylight"
                        >
                          12:18:32
                        </h2>,
                        "key": null,
                        "patchFlag": 1,
                        "props": Object {
                          "id": "daylight",
                        },
                        "ref": null,
                        "scopeId": null,
                        "shapeFlag": 9,
                        "slotScopeIds": null,
                        "ssContent": null,
                        "ssFallback": null,
                        "staticCount": 0,
                        "suspense": null,
                        "target": null,
                        "targetAnchor": null,
                        "transition": null,
                        "type": "h2",
                      },
                    ],
                    "component": null,
                    "dirs": null,
                    "dynamicChildren": null,
                    "dynamicProps": null,
                    "el": null,
                    "key": null,
                    "patchFlag": 0,
                    "props": Object {
                      "class": "py-1",
                    },
                    "ref": null,
                    "scopeId": null,
                    "shapeFlag": 17,
                    "slotScopeIds": null,
                    "ssContent": null,
                    "ssFallback": null,
                    "staticCount": 0,
                    "suspense": null,
                    "target": null,
                    "targetAnchor": null,
                    "transition": null,
                    "type": "div",
                  },
                ],
                "component": null,
                "dirs": null,
                "dynamicChildren": null,
                "dynamicProps": null,
                "el": null,
                "key": null,
                "patchFlag": 0,
                "props": Object {
                  "class": "px-6 py-3",
                },
                "ref": null,
                "scopeId": null,
                "shapeFlag": 17,
                "slotScopeIds": null,
                "ssContent": null,
                "ssFallback": null,
                "staticCount": 0,
                "suspense": null,
                "target": null,
                "targetAnchor": null,
                "transition": null,
                "type": "div",
              },
            ],
            "component": null,
            "dirs": null,
            "dynamicChildren": Array [
              Object {
                "__v_isVNode": true,
                "__v_skip": true,
                "anchor": null,
                "appContext": null,
                "children": "gio 9 giu 2022",
                "component": null,
                "dirs": null,
                "dynamicChildren": null,
                "dynamicProps": null,
                "el": <h3
                  id="date"
                >
                  gio 9 giu 2022
                </h3>,
                "key": null,
                "patchFlag": 1,
                "props": Object {
                  "id": "date",
                },
                "ref": null,
                "scopeId": null,
                "shapeFlag": 9,
                "slotScopeIds": null,
                "ssContent": null,
                "ssFallback": null,
                "staticCount": 0,
                "suspense": null,
                "target": null,
                "targetAnchor": null,
                "transition": null,
                "type": "h3",
              },
              Object {
                "__v_isVNode": true,
                "__v_skip": true,
                "anchor": null,
                "appContext": null,
                "children": "12:18:32",
                "component": null,
                "dirs": null,
                "dynamicChildren": null,
                "dynamicProps": null,
                "el": <h2
                  id="daylight"
                >
                  12:18:32
                </h2>,
                "key": null,
                "patchFlag": 1,
                "props": Object {
                  "id": "daylight",
                },
                "ref": null,
                "scopeId": null,
                "shapeFlag": 9,
                "slotScopeIds": null,
                "ssContent": null,
                "ssFallback": null,
                "staticCount": 0,
                "suspense": null,
                "target": null,
                "targetAnchor": null,
                "transition": null,
                "type": "h2",
              },
            ],
            "dynamicProps": null,
            "el": <main
              class="bg-white shadow-1 border-round flex-grow-1 p-2 m-1"
            >
              <h3
                id="date"
              >
                gio 9 giu 2022
              </h3>
              <div
                class="px-6 py-3"
              >
                <div
                  class="py-1"
                >
                  <small>
                    Day light:
                  </small>
                  <h2
                    id="daylight"
                  >
                    12:18:32
                  </h2>
                </div>
              </div>
            </main>,
            "key": null,
            "patchFlag": 0,
            "props": Object {
              "class": "bg-white shadow-1 border-round flex-grow-1 p-2 m-1",
            },
            "ref": null,
            "scopeId": null,
            "shapeFlag": 17,
            "slotScopeIds": null,
            "ssContent": null,
            "ssFallback": null,
            "staticCount": 0,
            "suspense": null,
            "target": null,
            "targetAnchor": null,
            "transition": null,
            "type": "main",
          },
          "suspense": null,
          "suspenseId": 0,
          "type": Object {
            "components": Object {},
            "created": [Function],
            "data": [Function],
            "methods": Object {
              "setDate": [Function],
              "setDaylight": [Function],
            },
            "props": Object {
              "cardDate": [Function],
            },
            "render": [Function],
          },
          "u": null,
          "uid": 1,
          "um": null,
          "update": [Function],
          "vnode": [Circular],
          "withProxy": null,
        },
        "dirs": null,
        "dynamicChildren": null,
        "dynamicProps": null,
        "el": <main
          class="bg-white shadow-1 border-round flex-grow-1 p-2 m-1"
        >
          <h3
            id="date"
          >
            gio 9 giu 2022
          </h3>
          <div
            class="px-6 py-3"
          >
            <div
              class="py-1"
            >
              <small>
                Day light:
              </small>
              <h2
                id="daylight"
              >
                12:18:32
              </h2>
            </div>
          </div>
        </main>,
        "key": null,
        "patchFlag": 0,
        "props": Object {
          "cardDate": "2022-06-09T00:00:00.000Z",
          "ref": "VTU_COMPONENT",
        },
        "ref": Object {
          "f": false,
          "i": [Circular],
          "k": undefined,
          "r": "VTU_COMPONENT",
        },
        "scopeId": null,
        "shapeFlag": 4,
        "slotScopeIds": null,
        "ssContent": null,
        "ssFallback": null,
        "staticCount": 0,
        "suspense": null,
        "target": null,
        "targetAnchor": null,
        "transition": null,
        "type": Object {
          "components": Object {},
          "created": [Function],
          "data": [Function],
          "methods": Object {
            "setDate": [Function],
            "setDaylight": [Function],
          },
          "props": Object {
            "cardDate": [Function],
          },
          "render": [Function],
        },
      },
      "suspense": null,
      "suspenseId": 0,
      "type": Object {
        "name": "VTU_ROOT",
        "render": [Function],
      },
      "u": null,
      "uid": 0,
      "um": null,
      "update": [Function],
      "vnode": Object {
        "__v_isVNode": true,
        "__v_skip": true,
        "anchor": null,
        "appContext": Object {
          "app": [Circular],
          "components": Object {
            "transition": Object {
              "name": "transition",
            },
            "transition-group": Object {
              "name": "transition-group",
            },
          },
          "config": Object {
            "compilerOptions": Object {},
            "errorHandler": undefined,
            "globalProperties": Object {},
            "isNativeTag": [Function],
            "optionMergeStrategies": Object {},
            "performance": false,
            "warnHandler": undefined,
          },
          "directives": Object {},
          "emitsCache": WeakMap {},
          "mixins": Array [
            Object {
              "beforeCreate": [Function],
            },
          ],
          "optionsCache": WeakMap {},
          "propsCache": WeakMap {},
          "provides": Object {},
          "reload": [Function],
        },
        "children": null,
        "component": [Circular],
        "dirs": null,
        "dynamicChildren": null,
        "dynamicProps": null,
        "el": <main
          class="bg-white shadow-1 border-round flex-grow-1 p-2 m-1"
        >
          <h3
            id="date"
          >
            gio 9 giu 2022
          </h3>
          <div
            class="px-6 py-3"
          >
            <div
              class="py-1"
            >
              <small>
                Day light:
              </small>
              <h2
                id="daylight"
              >
                12:18:32
              </h2>
            </div>
          </div>
        </main>,
        "key": null,
        "patchFlag": 0,
        "props": null,
        "ref": null,
        "scopeId": null,
        "shapeFlag": 4,
        "slotScopeIds": null,
        "ssContent": null,
        "ssFallback": null,
        "staticCount": 0,
        "suspense": null,
        "target": null,
        "targetAnchor": null,
        "transition": null,
        "type": Object {
          "name": "VTU_ROOT",
          "render": [Function],
        },
      },
      "withProxy": null,
    },
    "_props": null,
    "_uid": 0,
    "component": [Function],
    "config": Object {
      "compilerOptions": Object {},
      "errorHandler": undefined,
      "globalProperties": Object {},
      "isNativeTag": [Function],
      "optionMergeStrategies": Object {},
      "performance": false,
      "warnHandler": undefined,
    },
    "directive": [Function],
    "mixin": [Function],
    "mount": [Function],
    "provide": [Function],
    "unmount": [Function],
    "use": [Function],
    "version": "3.2.39",
  },
  "__setProps": [Function],
  "componentVM": Object {
    "cardDate": "2022-06-09T00:00:00.000Z",
    "date": "gio 9 giu 2022",
    "daylight": "12:18:32",
    "hasOwnProperty": [Function],
    "setDate": [Function],
    "setDaylight": [Function],
  },
  "isDisabled": [Function],
  "rootVM": Object {},
  "wrapperElement": <main
    class="bg-white shadow-1 border-round flex-grow-1 p-2 m-1"
  >
    <h3
      id="date"
    >
      gio 9 giu 2022
    </h3>
    <div
      class="px-6 py-3"
    >
      <div
        class="py-1"
      >
        <small>
          Day light:
        </small>
        <h2
          id="daylight"
        >
          12:18:32
        </h2>
      </div>
    </div>
  </main>,
}
`;
